src/nextjournal/clerk/analyzer.clj

```diff
@@ -234,39 +253,46 @@
   (let [rec-form (concat '(do) [form (get-in state [:->analysis-info dependency :form])])
         rec-var (symbol (str/join "+" (sort (conj vars dep))))
         var (first vars)] ;; TODO: reduce below
+    ;;(when (not= dep dependency) (println :dep-mismatch dep dependency))
+    ;;(when (not= var node) (println :node-mismatch var node))
     (-> state
         (update :graph #(-> %
                             (dep/remove-edge dependency node)
                             (dep/depend var rec-var)
                             (dep/depend dep rec-var)))
         (assoc-in [:->analysis-info rec-var :form] rec-form))))
 
+#_(->> (nextjournal.clerk.eval/eval-string "(rand-int 100) (rand-int 100) (rand-int 100)") :blocks (mapv #(-> % :result :nextjournal/value)))
 
-(defn ->key [{:as codeblock :keys [var id]}]
-  (if var var id))
-
-(defn ->ana-keys [{:as analyzed :keys [form vars id]}]
-  (if (seq vars) vars [(->key analyzed)]))
+(defn dep->block-id
+  "Inverse key-lookup from deps to block ids"
+  [{:keys [var->block-id]} dep]
+  (or (get var->block-id dep) dep))
 
-#_(->> (nextjournal.clerk.eval/eval-string "(rand-int 100) (rand-int 100) (rand-int 100)") :blocks (mapv #(-> % :result :nextjournal/value)))
+(defn deps->block-ids [state {:as _info :keys [deps]}]
+  (into #{} (map (partial dep->block-id state)) deps))
 
-(defn- analyze-deps [{:as analyzed :keys [form vars]} state dep]
-  (try (reduce (fn [state _var] ;; TODO: check if `_var` needs to be used
-                 (update state :graph #(dep/depend % (->key analyzed) dep)))
-               state
-               (->ana-keys analyzed))
+(defn- analyze-deps [{:as _info :keys [id form vars]} state dep]
+  (try (update state :graph dep/depend id (dep->block-id state dep))
        (catch Exception e
          (if (circular-dependency-error? e)
            (analyze-circular-dependency state vars form dep (ex-data e))
            (throw e)))))
 
-
-(defn make-deps-inherit-no-cache [state {:as analyzed :keys [no-cache? vars deps ns-effect?]}]
-  (if no-cache?
-    state
-    (let [no-cache-deps? (boolean (some (fn [dep] (get-in state [:->analysis-info dep :no-cache?])) deps))]
-      (reduce (fn [state k]
-                (assoc-in state [:->analysis-info k :no-cache?] no-cache-deps?)) state (->ana-keys analyzed)))))
+(defn make-deps-inherit-no-cache [{:as doc :keys [ns ->analysis-info blocks]}]
+  (let [nsname (when ns (str (ns-name ns)))]
+    (reduce (fn no-cache-inheritance-rf [state {:as analyzed :keys [id no-cache?]}]
+              (if no-cache?
+                state
+                (assoc-in state [:->analysis-info id :no-cache?]
+                          (boolean (some (fn [k]
+                                           (and ns (symbol? k) (= nsname (namespace k))
+                                                (get-in state [:->analysis-info k :no-cache?])))
+                                         (deps->block-ids state analyzed))))))
```

Comment:
after discussion on #633, we now _ensure_ that `:no-cache?` does not propagate across files.